### PR TITLE
Add APM Single Step Instrumentation

### PR DESCRIPTION
**Pull request by CharlyF**
### What does this PR do?

Define a new struct in the APM feature to allow configuration of the Single Instrumentation Step. Sister PR of https://github.com/DataDog/helm-charts/pull/1211.
Added a couple of tests, the CRD change and the main code in distinct commits.

### Motivation

To be confirmed if the following is the expected format.
```
    - name: DD_APM_INSTRUMENTATION_ENABLED
      value: "true"
    - name: DD_APM_INSTRUMENTATION_DISABLED_NAMESPACES
      value: '["foo"]'
    - name: DD_APM_INSTRUMENTATION_LIB_VERSIONS
      value: '{"java":"1.2.3","test":"1.0.0"}'
```
but the DCA's config seems good:
```
  instrumentation:
    disabled_namespaces:
    - foo
    enabled: true
    enabled_namespaces: []
    lib_versions: '{"java":"1.2.3","test":"1.0.0"}'
```

### Additional Notes

Anything else we should know when reviewing?

### Minimum Agent Versions

Are there minimum versions of the Datadog Agent and/or Cluster Agent required?

* Agent: N/A
* Cluster Agent: v7.39.0

### Describe your test plan

```
spec:
  features:
    apm:
      enabled: true
      singleStepInstrumentation:
        disabledNamespaces:
        - foo
        enabled: true
        libVersions:
          java: 1.2.3
          test: 1.0.0
```
should start the Cluster Agent and configure: `DD_APM_INSTRUMENTATION_ENABLED`,  `DD_APM_INSTRUMENTATION_DISABLED_NAMESPACES` and `DD_APM_INSTRUMENTATION_LIB_VERSIONS`. 
If you use `DD_APM_INSTRUMENTATION_DISABLED_NAMESPACES` and `DD_APM_INSTRUMENTATION_ENABLED_NAMESPACES`, it will error out (not supported for now?).

```
{"level":"ERROR","ts":"2023-11-15T03:50:36Z","logger":"controller.datadogagent","msg":"Reconciler error","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","name":"datadog","namespace":"default","error":"`enabledMamespaces` and `disabledNamespaces` cannot be set together"}
```

### Checklist

- [x] PR has at least one valid label: `bug`, `enhancement`, `refactoring`, `documentation`, `tooling`, and/or `dependencies`
- [x] PR has a milestone or the `qa/skip-qa` label

### CODE REVIEW

**Comment by liliyadd** [comment]
Thanks @CharlyF for the PR!

In both helm and Cluster Agent, we have some validation of SSI parameters. I was wondering if https://github.com/DataDog/datadog-operator/blob/main/controllers/datadogagent/clusteragent.go#L689 would be the right place to put it for the Operator.

The validation logic is the following:
```
1. If spec.features.enabledNamespaces != nil AND spec.features.disabledNamespaces != nil, return the error with the message that enabledNamespaces and disabledNamespaces cannot be set at the same time. Datadog Operator should fail installing in this case.
2. If spec.features.enabled='true' AND spec.features.enabledNamespaces != nil, show the warning saying SSI will be enabled in the whole cluster.
3.  If spec.features.enabled='false' AND spec.features.disabledNamespaces != nil, show the warning saying SSI will be disabled in the whole cluster.
```

**Comment by liliyadd** [comment]
Oh I see this logic is in features.go. Will throwing an error there fail Operator installation?

**Review by hestonhoffman — APPROVED** [review]

File: controllers/datadogagent/feature/apm/feature.go [hunk 1]
```diff
@@ -228,6 +252,47 @@ func (f *apmFeature) ManageDependencies(managers feature.ResourceManagers, compo
 // ManageClusterAgent allows a feature to configure the ClusterAgent's corev1.PodTemplateSpec
 // It should do nothing if the feature doesn't need to configure it.
 func (f *apmFeature) ManageClusterAgent(managers feature.PodTemplateManagers) error {
+	if f.singleStepInstrumentation != nil && f.singleStepInstrumentation.enabled {
```
[the remark this answers is not in the document]
Comment (hkaj) [reply controllers/datadogagent/feature/apm/feature.go:255]: hmm that would be super confusing UX imo. The flag explicitly says `enabled: false` but then it's not actually disabled? Why should we do that?

[the remark this answers is not in the document]
Comment (liliyadd) [reply controllers/datadogagent/feature/apm/feature.go:255]: `singleStepInstrumentation.enabled` is a cluster level configuration, `singleStepInstrumentation.enabledNamespace` is a namespace level configuration. SSI can be disabled in the cluster, but enabled on a specific namespace.

[the remark this answers is not in the document]
Comment (hkaj) [reply controllers/datadogagent/feature/apm/feature.go:255]: Let's not do that please. This would be unique behavior, no other product does that. It's also confusing, as a user you'd think you're disabling the feature but it's not actually the case. Global enable/disable should be global, namespace-level config should be namespace level, etc. and global should override everything, that's how every other config of the agent works.

File: docs/configuration.v2alpha1.md [hunk 2]
```diff
@@ -43,6 +43,10 @@ spec:
 | features.apm.enabled | Enabled enables Application Performance Monitoring. Default: false |
```
Comment (fanny-jiang) [inline docs/configuration.v2alpha1.md:43]: nit

```suggestion
| features.apm.enabled | Enabled enables Application Performance Monitoring. Default: true |
```

**Review by fanny-jiang — CHANGES_REQUESTED** [review]

File: controllers/datadogagent/feature/apm/feature.go [hunk 3]
```diff
@@ -228,6 +266,48 @@ func (f *apmFeature) ManageDependencies(managers feature.ResourceManagers, compo
 // ManageClusterAgent allows a feature to configure the ClusterAgent's corev1.PodTemplateSpec
 // It should do nothing if the feature doesn't need to configure it.
 func (f *apmFeature) ManageClusterAgent(managers feature.PodTemplateManagers) error {
+	if f.singleStepInstrumentation != nil {
+		if len(f.singleStepInstrumentation.disabledNamespaces) > 0 && len(f.singleStepInstrumentation.enabledNamespaces) > 0 {
+			// This configuration is not supported
+			return fmt.Errorf("`enabledMamespaces` and `disabledNamespaces` cannot be set together")
```
Comment (levan-m) [inline controllers/datadogagent/feature/apm/feature.go:272]: nit - typo in `enabledMamespaces`

File: controllers/datadogagent/feature/apm/feature.go [hunk 4]
```diff
@@ -54,18 +56,39 @@ type apmFeature struct {
...
 func (f *apmFeature) ID() feature.IDType {
 	return feature.APMIDType
 }
 
+func shouldEnableAPM(apmConf *v2alpha1.APMFeatureConfig) bool {
+	if apmConf != nil {
+		return apiutils.BoolValue(apmConf.Enabled)
+	}
+	// SingleStepInstrumentation requires APM Enabled
+	if apmConf.SingleStepInstrumentation != nil &&
```
Comment (levan-m) [inline controllers/datadogagent/feature/apm/feature.go:80]: if `apmConf.Enabled == false` function will return `false` and SSI settings don't matter, otherwise it will return `true`. Only time this will be evaluated is when `apmConf == nil`. 

If `apmConf.Enabled == true` and `apmConf.SSI == nil` this still should return true right?

I think having a unit test describe these expectations would be beneficial.

File: controllers/datadogagent/feature/apm/feature_test.go [hunk 5]
```diff
@@ -229,6 +326,123 @@ func testAgentUDSOnly() *test.ComponentTest {
...
+			assert.True(
+				t,
+				apiutils.IsEqualStruct(agentEnvs, expectedAgentEnvs),
+				"Cluster Agent ENVs \ndiff = %s", cmp.Diff(agentEnvs, expectedAgentEnvs),
+			)
+		},
+	)
+}
+
+func testAPMInstrumentationPrecedence() *test.ComponentTest {
```
Comment (levan-m) [inline controllers/datadogagent/feature/apm/feature_test.go:404]: nit - unused

File: controllers/datadogagent/feature/apm/feature.go (hunk 4 above)
Comment (CharlyF) [reply controllers/datadogagent/feature/apm/feature.go:80]: Yes that's correct. Essentially, this tests that APM enabled by default AND that it takes precedence over SSI if configured. 
I am adding a test to reflect the `apmConf == nil` behavior, but:
> If apmConf.Enabled == true and apmConf.SSI == nil this still should return true right?
is tested with `v2alpha1 apm enabled, use uds` and `v2alpha1 apm enabled, use uds and host port`.

File: apis/datadoghq/v2alpha1/datadogagent_default_test.go [hunk 6]
```diff
@@ -129,6 +129,17 @@ func Test_defaultFeatures(t *testing.T) {
...
+						Enabled: apiutils.NewBoolPointer(defaultAPMEnabled),
+						HostPortConfig: &HostPortConfig{
+							Port:    apiutils.NewInt32Pointer(defaultAPMHostPort),
+							Enabled: apiutils.NewBoolPointer(defaultAPMHostPortEnabled),
+						},
+						UnixDomainSocketConfig: &UnixDomainSocketConfig{
+							Enabled: apiutils.NewBoolPointer(defaultAPMSocketEnabled),
+							Path:    apiutils.NewStringPointer(defaultAPMSocketHostPath),
+						},
+					},
```
Comment (liliyadd) [inline apis/datadoghq/v2alpha1/datadogagent_default_test.go:142]: Do we need to add defaults for `apm.instrumentation` block?

Comment (CharlyF) [reply apis/datadoghq/v2alpha1/datadogagent_default_test.go:142]: Yep, added.

File: controllers/datadogagent/feature/apm/feature.go [hunk 7]
```diff
@@ -228,6 +266,48 @@ func (f *apmFeature) ManageDependencies(managers feature.ResourceManagers, compo
 // ManageClusterAgent allows a feature to configure the ClusterAgent's corev1.PodTemplateSpec
 // It should do nothing if the feature doesn't need to configure it.
 func (f *apmFeature) ManageClusterAgent(managers feature.PodTemplateManagers) error {
+	if f.singleStepInstrumentation != nil {
+		if len(f.singleStepInstrumentation.disabledNamespaces) > 0 && len(f.singleStepInstrumentation.enabledNamespaces) > 0 {
+			// This configuration is not supported
+			return fmt.Errorf("`enabledNamespaces` and `disabledNamespaces` cannot be set together")
```
Comment (levan-m) [inline controllers/datadogagent/feature/apm/feature.go:272]: Behavior I saw when testing https://github.com/DataDog/helm-charts/pull/1252 was not to error and instead use the list based on the flag selection - i.e. if enabled use `disabledNamespaces`, if disabled use `enabledNamespaces`. Returning error will prevent reconciling and any agent installation.

Could you confirm this is what we want, @liliyadd?

File: docs/configuration.v2alpha1.md [hunk 8]
```diff
@@ -40,9 +40,13 @@ spec:
 | features.admissionController.mutateUnlabelled | MutateUnlabelled enables config injection without the need of pod label 'admission.datadoghq.com/enabled="true"'. Default: false |
 | features.admissionController.serviceName | ServiceName corresponds to the webhook service name. |
 | features.admissionController.webhookName | WebhookName is a custom name for the MutatingWebhookConfiguration. Default: "datadog-webhook" |
-| features.apm.enabled | Enabled enables Application Performance Monitoring. Default: false |
+| features.apm.enabled | Enabled enables Application Performance Monitoring. Default: true |
 | features.apm.hostPortConfig.enabled | Enabled enables host port configuration Default: false |
 | features.apm.hostPortConfig.hostPort | Port takes a port number (0 < x < 65536) to expose on the host. (Most containers do not need this.) If HostNetwork is enabled, this value must match the ContainerPort. |
+| features.apm.singleStepInstrumentation.disabledNamespaces | DisabledNamespaces specifies the namespaces of the pods to not inject the Datadog APM libraries into. |
```
Comment (liliyadd) [inline docs/configuration.v2alpha1.md:46]: Can we please change the configuration names to:
`features.apm.instrumentation.enabled`
`features.apm.instrumentation.enabledNamespaces`
`features.apm.instrumentation.disabledNamespaces`
`features.apm.instrumentation.libVersions`

These names will align with the naming used in Helm Charts.

**Review by fanny-jiang — APPROVED** [review]

**Review by levan-m — APPROVED** [review]

**Review by liliyadd — APPROVED** [review]
The changes in the PR look good to me. There are some discrepancies between  Operator and current Helm chart behavior around disabling SSI in a Datadog namespace by default that @CharlyF and I discussed - Operator won't support disabling SSI in DD namespace and this behavior needs to be documented for the end user. The logic of disabling SSI in DD namespace will be moved to Cluster Agent in 7.51 Agent release allowing Helm Chart cleanup and making Helm and Operator consistent.
